src/UsgsAstroPlugin.cpp

```diff
@@ -0,0 +1,306 @@
+const int         UsgsAstroPlugin::_NUM_ISD_KEYWORDS = 21;
+
+const std::string UsgsAstroPlugin::_ISD_KEYWORD[] =
+{
+   "model_name",
+   "center_ephemeris_time",
+   "dt_ephemeris",
+   "focal2pixel_lines",
+   "focal2pixel_samples",
+   "focal_length_model",
+   "image_lines",
+   "image_samples",
+   "interpolation_method",
+   "number_of_ephemerides",
+   "optical_distortion",
+   "radii",
+   "reference_height",
+   "sensor_location_unit",
+   "sensor_location",
+   "sensor_orientation",
+   "sensor_velocity",
+   "detector_center",
+   "starting_detector_line",
+   "starting_detector_sample",
+   "starting_ephemeris_time",
+   "sun_position"
+};
+
+// const json UsgsAstroPlugin::MODEL_KEYWORDS = {
+//   {UsgsAstroFrameSensorModel::_SENSOR_MODEL_NAME, UsgsAstroFrameSensorModel::_STATE_KEYWORD},
+//   {UsgsAstroLsSensorModel::_SENSOR_MODEL_NAME, UsgsAstroLsSensorModel::_STATE_KEYWORD}
+// };
+
+
+// Static Instance of itself
+const UsgsAstroPlugin UsgsAstroPlugin::m_registeredPlugin;
+
+UsgsAstroPlugin::UsgsAstroPlugin() {
+}
+
+
+UsgsAstroPlugin::~UsgsAstroPlugin() {
+}
+
+
+std::string UsgsAstroPlugin::getPluginName() const {
+  return _PLUGIN_NAME;
+}
+
+
+std::string UsgsAstroPlugin::getManufacturer() const {
+  return _MANUFACTURER_NAME;
+}
+
+
+std::string UsgsAstroPlugin::getReleaseDate() const {
+  return _RELEASE_DATE;
+}
+
+
+csm::Version UsgsAstroPlugin::getCsmVersion() const {
+  return CURRENT_CSM_VERSION;
+}
+
+
+size_t UsgsAstroPlugin::getNumModels() const {
+  return _N_SENSOR_MODELS;
+}
+
+
+std::string UsgsAstroPlugin::getModelName(size_t modelIndex) const {
+  // return SUPPORTED_MODELS[modelIndex];
+  return "";
+}
+
+
+std::string UsgsAstroPlugin::getModelFamily(size_t modelIndex) const {
+  return CSM_RASTER_FAMILY;
+}
+
+
+csm::Version UsgsAstroPlugin::getModelVersion(const std::string &modelName) const {
+  return csm::Version(1, 0, 0);
+}
+
+
+bool UsgsAstroPlugin::canModelBeConstructedFromState(const std::string &modelName,
+                                                const std::string &modelState,
+                                                csm::WarningList *warnings) const {
+
+  bool constructible = true;
+
+  // Get the model name from the model state
+  std::string model_name_from_state;
+  try {
+    model_name_from_state = getModelNameFromModelState(modelState, warnings);
+  }
+  catch(...) {
+    return false;
+  }
+
+  // Check that the plugin supports the model
+  if (modelName != model_name_from_state ||
+      modelName != UsgsAstroFrameSensorModel::_SENSOR_MODEL_NAME){
+          constructible = false;
+      }
+  // Check that the necessary keys are there (this does not chek values at all.)
+  // This sort of check should probably be moved into the sensor model
+  // auto state = json::parse(modelState);
+  // for(auto &key : MODEL_KEYWORDS[model_name_from_state]){
+  //     if (state.find(key) == state.end()){
+  //         constructible = false;
+  //         break;
+  //     }
+  // }
```

Comment:
I'm worried that if we don't check this here, some SETs will fail. They probably expect a false from this call and then will not call constructFromState. If they get a true from this, and then an exception from constructFromState they will probably completely fail.

Reply:
Right now, the sensor model checks that the state has all of the required values in the replaceModelState function. I would move all of that into another function that would return a list of missing/invalid keywords. Then, in replaceModelState call that function to check the state is good. The plugin, can then call that same function here.